cmd/kubeapps-apis/server/plugins.go

```diff
@@ -232,3 +253,140 @@ func listSOFiles(fsys fs.FS, pluginDirs []string) ([]string, error) {
+	}
+
+	configs := kube.ClustersConfig{Clusters: map[string]kube.ClusterConfig{}}
+	configs.PinnipedProxyURL = pinnipedProxyURL
+	for _, c := range clusterConfigs {
+		// Select the cluster in which Kubeapps in installed. We look for either
+		// `isKubeappsCluster: true` or an empty `APIServiceURL`.
+		isKubeappsClusterCandidate := c.IsKubeappsCluster || c.APIServiceURL == ""
+		if isKubeappsClusterCandidate {
+			if configs.KubeappsClusterName == "" {
+				configs.KubeappsClusterName = c.Name
+			} else {
+				return kube.ClustersConfig{}, nil, fmt.Errorf("only one cluster can be configured using either 'isKubeappsCluster: true' or without an apiServiceURL to refer to the cluster on which Kubeapps is installed, two defined: %q, %q", configs.KubeappsClusterName, c.Name)
+			}
+		}
+
+		// We need to decode the base64-encoded cadata from the input.
+		if c.CertificateAuthorityData != "" {
+			decodedCAData, err := base64.StdEncoding.DecodeString(c.CertificateAuthorityData)
+			if err != nil {
+				return kube.ClustersConfig{}, deferFn, err
+			}
+			c.CertificateAuthorityDataDecoded = string(decodedCAData)
+
+			// We also need a CAFile field because Helm uses the genericclioptions.ConfigFlags
+			// struct which does not support CAData.
+			// https://github.com/kubernetes/cli-runtime/issues/8
+			c.CAFile = filepath.Join(caFilesDir, c.Name)
+			err = ioutil.WriteFile(c.CAFile, decodedCAData, 0644)
+			if err != nil {
+				return kube.ClustersConfig{}, deferFn, err
+			}
+		}
+		configs.Clusters[c.Name] = c
+	}
+	return configs, deferFn, nil
+}
+
+// dynClientGetterForContext returns a k8s client for use during interactions with the cluster.
+// It is invoked by dynClientGetterForContext and unit tests passing the appropriate configuration
+func dynClientGetterForContextWithConfig(ctx context.Context, inClusterConfig *rest.Config, serveOpts ServeOptions, config kube.ClustersConfig) (dynamic.Interface, error) {
+	var err error
+	token, err := extractToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var client dynamic.Interface
+	if !serveOpts.UnsafeUseDemoSA {
+		restConfig, err := kube.NewClusterConfig(inClusterConfig, token, "default", config)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get clusterConfig: %w", err)
+		}
+		client, err = dynamic.NewForConfig(restConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create dynamic client: %w", err)
+		}
+	} else {
+		client, err = dynamic.NewForConfig(inClusterConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create dynamic client: %w", err)
+		}
+	}
+	return client, nil
+}
+
+// dynClientGetterForContext returns a k8s client for use during interactions with the cluster.
+// It utilizes the user credential from the request context. The plugins just have to call this function
+// passing the context in order to retrieve the configured k8s client
+func dynClientGetterForContext(ctx context.Context) (dynamic.Interface, error) {
```

Comment:
As above, I think this is the reason you've needed global vars - so you can reference `pluginsServeOpts` inside this static function. But using a function closure should mean you don't need the global var at all.